### PR TITLE
[Merged by Bors] - chore: better error handling in congr!

### PR DESCRIPTION
**Pull request by kim-em**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: test/convert2.lean [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+import Mathlib.Data.Nat.Interval
+
+open BigOperators
+
+-- Prior to #7945 this failed with `(kernel) declaration has metavariables '_example'`.
+-- There is in fact an internal error being suppressed here, that is reported only with
+-- `set_option trace.congr! true`
+example (h : False) (_h₁ : ∑ k in Finset.Icc 0 122, 0 = 0) : 0 ∣ 1 := by
+  apply Nat.dvd_of_mul_dvd_mul_left zero_lt_one
+  convert dvd_mul_left 0 1
```
Comment (fpvandoorn) [inline test/convert2.lean:12]: I think this tactic should raise an error that the discharger times out, see [my message here](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/bug.20in.20convert/near/398700746)

File: Mathlib/Tactic/Congr!.lean [hunk 2]
```diff
@@ -618,9 +618,20 @@ def Lean.MVarId.congrN! (mvarId : MVarId)
...
+        -- but it is somehow(!?) caught, and the mvarId is lost as a consequence.
+        -- throwTacticEx `congr! mvarId m!"internal error in postCongr!, {e.toMessageData}"
+        trace[congr!] "internal error in postCongr!, {e.toMessageData}"
+        throw e
+      match mvarId? with
+      | some mvarId =>
+        trace[congr!] "Recording {mvarId}."
         modify (fun s => {s with goals := s.goals.push mvarId})
-      else
+      | none =>
```
Comment (kmill) [inline Mathlib/Tactic/Congr!.lean:634]: Given Eric's fix to Convert.lean, do we need to make this change anymore? If we want logging, I think I'd rather that `postCongr!` itself have its body wrapped in a `try` block, where it re-throws the error but mentions that the error is passing through `postCongr!`.

File: test/convert2.lean [hunk 3]
```diff
@@ -0,0 +1,10 @@
+import Mathlib.Data.List.Defs
+import Mathlib.Data.Nat.Basic
+
+-- Prior to #7945 this failed with `(kernel) declaration has metavariables '_example'`.
+-- There is in fact an internal error being suppressed here, that is reported only with
+-- `set_option trace.congr! true`
+example (h : False) (_h₁ : ((List.range 128).map (fun _ => 0)).sum = 0) : 0 ∣ 1 := by
+  apply Nat.dvd_of_mul_dvd_mul_left Nat.zero_lt_one
+  convert Nat.dvd_mul_left 0 1
```
Comment (eric-wieser) [inline test/convert2.lean:9]: This now throws an error, but I don't remember what the preferred way to test it is.

File: Mathlib/Tactic/Congr!.lean (hunk 2 above)
Comment (kmill) [reply Mathlib/Tactic/Congr!.lean:634]: There's `Lean.Meta.throwNestedTacticEx` (check its implementation to see how simple it is -- if we want better context, then may as well just `throwError` like it does).

There's also the fancier `Lean.Elab.throwErrorWithNestedErrors`, which seems to actually report source positions of the nested errors. Mathlib never uses it directly, and instead it uses `Lean.Elab.nestedExceptionToMessageData` (see `focusAndDoneWithScope`)

File: test/convert2.lean (hunk 3 above)
Comment (kmill) [reply test/convert2.lean:9]: You could do `#guard_msgs (error) in example ...` to just capture the error message.

**Pull request closed without merging**